editor/components/inserter/menu.js

```diff
@@ -29,32 +77,206 @@ class InserterMenu extends wp.element.Component {
 		return () => {
 			this.props.onInsertBlock( slug );
 			this.props.onSelect();
-			this.setState( { filterValue: '' } );
+			this.setState( {
+				filterValue: '',
+				currentFocus: null
+			} );
 		};
 	}
 
-	render() {
-		const { position = 'top' } = this.props;
-		const blocks = wp.blocks.getBlocks();
-		const isShownBlock = block => block.title.toLowerCase().indexOf( this.state.filterValue.toLowerCase() ) !== -1;
-		const blocksByCategory = blocks.reduce( ( groups, block ) => {
-			if ( ! isShownBlock( block ) ) {
-				return groups;
-			}
-			if ( ! groups[ block.category ] ) {
-				groups[ block.category ] = [];
+	getVisibleBlocks( blockTypes ) {
+		return blockTypes.filter( this.isShownBlock );
+	}
+
+	sortBlocksByCategory( blockTypes ) {
+		const getCategoryIndex = ( item ) => {
+			return this.categories.findIndex( ( category ) => category.slug === item.slug );
+		};
+
+		return blockTypes.sort( ( a, b ) => {
+			return getCategoryIndex( a ) - getCategoryIndex( b );
+		} );
+	}
+
+	groupByCategory( blockTypes ) {
+		return blockTypes.reduce( ( accumulator, block ) => {
+			// If already an array push block on else add array of block.
+			if ( Array.isArray( accumulator[ block.category ] ) ) {
+				accumulator[ block.category ].push( block );
+				return accumulator;
 			}
-			groups[ block.category ].push( block );
-			return groups;
+
+			accumulator[ block.category ] = [ block ];
+			return accumulator;
 		}, {} );
-		const categories = wp.blocks.getCategories();
+	}
+
+	getVisibleBlocksByCategory( blockTypes ) {
+		return flow(
+			this.getVisibleBlocks,
+			this.sortBlocksByCategory,
+			this.groupByCategory
+		)( blockTypes );
+	}
+
+	findNext( currentBlock, blockTypes ) {
+		/**
+		 * null is the value that will trigger iterating back to
+		 * the top of the list of block types.
+		 */
+		if ( null === currentBlock ) {
+			return blockTypes[ 0 ].slug;
+		}
+
+		const currentIndex = blockTypes.findIndex( ( blockType ) => currentBlock === blockType.slug );
+		const nextIndex = currentIndex + 1;
+		const highestIndex = blockTypes.length - 1;
+
+		/**
+		 * Default currently for going past the blocks is search, may need to be
+		 * revised in the future as more focusable elements are added. This
+		 * returns a null value, which currently implies that search will be set
+		 * as the next focus.
+		 */
+		if ( nextIndex > highestIndex ) {
+			return null;
+		}
+
+		// Return the slug of the next block type.
+		return blockTypes[ nextIndex ].slug;
+	}
+
+	findPrevious( currentBlock, blockTypes ) {
+		/**
```

Comment:
There's a fair bit of overlap between `findNext` and `findPrevious` implementations. I'm wondering if it could be consolidated into a single one, or at least a single base implementation with aliases. Something like:

```
findByIncrement( increment = 1 ) {
```